torch/distributed/_composable/fsdp/_fsdp_collectives.py

```diff
@@ -14,6 +14,25 @@ class AllGatherResult(NamedTuple):
     all_gather_input_numels: List[int]
 
 
+class AllGatherState(NamedTuple):
+    all_gather_result: AllGatherResult
+    event: torch.cuda.Event  # copy-out
```

Comment:
The `AllGatherState` is shared across `FSDPParamGroup`s to implement overlapping copy-in with all-gather. We need to keep the current all-gather result alive until we copy-in for the next all-gather.

Because this is some shared object across `FSDPParamGroup`s, I had it as a separate class. Alternatively, we could implement it as something like a singleton list, but I felt that that might be less clear.

Reply:
I deleted the original comment to try to re-explain.

This `AllGatherState` and `AllGatherStateHolder` are used to implement overlapping the next all-gather copy-in with the current all-gather. To do so, we must keep the current all-gather result alive until that copy-in finishes to avoid reusing the memory. This is why this `AllGatherState` has an `all_gather_result` field but never uses it. It is only for the reference.

The `AllGatherStateHolder` object needs to be shared across all `FSDPParamGroup`s as the next parameter group is responsible for freeing the current's all-gather result.

Reply:
I am working on refactoring this in https://github.com/pytorch/pytorch/pull/118004.